crates/guest-rust/rt/src/async_support/waitable.rs

```diff
@@ -0,0 +1,377 @@
+struct CompletionStatus {
+    /// Where the async operation's code is filled in, and `None` until that
+    /// happens.
+    code: Option<u32>,
+
+    /// This is necessary to ensure that `Pin<&mut CompletionStatus>` carries
+    /// the "pin guarantee", basically to mean that it's not safe to construct
+    /// `Pin<&mut CompletionStatus>` and it must somehow require `unsafe` code.
+    _pinned: marker::PhantomPinned,
+}
+
+/// Helper trait to be used with `WaitableOperation` to assist with machinery
+/// necessary to track in-flight reads/writes on futures.
+///
+/// # Unsafety
+///
+/// This trait is `unsafe` as it has various guarantees that must be upheld by
+/// implementors such as:
+///
+/// * `S::in_progress_waitable` must always return the same value for the state
+///   given.
+pub unsafe trait WaitableOp {
+    /// Initial state of this operation, used to kick off the actual component
+    /// model operation and transition to `InProgress`.
+    type Start;
+
+    /// Intermediate state of this operation when the component model is
+    /// involved but it hasn't resolved just yet.
+    type InProgress;
+
+    /// Result type of this operation.
+    type Result;
+
+    /// Result of when this operation is cancelled.
+    type Cancel;
+
+    /// Starts the async operation.
+    ///
+    /// This method will actually call `{future,stream}.{read,write}` with
+    /// `state` provided. The return code of the intrinsic is returned here
+    /// along with the `InProgress` state.
+    fn start(state: Self::Start) -> (u32, Self::InProgress);
+
+    /// Conversion from the "start" state to the "cancel" result, needed when an
+    /// operation is cancelled before it's started.
+    fn start_cancel(state: Self::Start) -> Self::Cancel;
+
+    /// Completion callback for when an in-progress operation has completed
+    /// successfully after transferring `amt` items.
+    fn in_progress_complete(state: Self::InProgress, amt: u32) -> Self::Result;
+
+    /// Completion callback for when an in-progress operation has completed
+    /// without actually transferring anything because the other end has closed.
+    fn in_progress_closed(state: Self::InProgress) -> Self::Result;
+
+    /// Acquires the component-model `waitable` index that the `InProgress`
+    /// state is waiting on.
+    fn in_progress_waitable(state: &Self::InProgress) -> u32;
+
+    /// Initiates a request for cancellation of this operation. Returns the
+    /// status code returned by the `{future,stream}.cancel-{read,write}`
+    /// intrinsic.
+    fn in_progress_cancel(state: &Self::InProgress) -> u32;
+
+    /// Completion callback for when an operation was canceled.
+    ///
+    /// This is invoked after `in_progress_cancel` is used and the returned
+    /// status code indicates that the operation was indeed cancelled and didn't
+    /// racily return some other result.
+    fn in_progress_canceled(state: Self::InProgress) -> Self::Result;
```

Comment:
Some comments and code spell it `cancelled` and some spell it `canceled`.  Apparently [either is correct](https://www.merriam-webster.com/grammar/canceled-or-cancelled), but we should probably choose one and use it consistently.